rasa/core/lock_store.py

```diff
@@ -0,0 +1,262 @@
+            )
+            lock_store = LockStore.load_lock_store_from_module_path(store.type)
+
+        logger.debug(
+            "Connected to lock store '{}'.".format(lock_store.__class__.__name__)
+        )
+
+        return lock_store
+
+    @staticmethod
+    def load_lock_store_from_module_path(module_path: Text) -> "LockStore":
+        """Given the name of a `LockStore` module tries to retrieve it."""
+
+        from rasa.utils.common import class_from_module_path
+
+        try:
+            return class_from_module_path(module_path)
+        except ImportError:
+            raise ImportError(
+                "Cannot retrieve `LockStore` from path '{}'.".format(module_path)
+            )
+
+    def create_lock(self, conversation_id: Text) -> TicketLock:
+        """Create and save a new `TicketLock` for `conversation_id`."""
+
+        lock = TicketLock(conversation_id)
+        self.save_lock(lock)
+        return lock
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        """Fetch lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        """Delete lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def save_lock(self, lock: TicketLock) -> None:
+        """Commit `lock` to storage."""
+
+        raise NotImplementedError
+
+    def issue_ticket(
+        self, conversation_id: Text, lifetime: Union[float, int] = DEFAULT_LOCK_LIFETIME
+    ) -> int:
+        """Issue new ticket for lock associated with `conversation_id`.
+
+        Creates a new lock if none is found.
+        """
+
+        lock = self.get_or_create_lock(conversation_id)
+
+        lifetime = lifetime if lifetime is not None else self.lifetime
+        ticket = lock.issue_ticket(lifetime)
+
+        self.save_lock(lock)
+        return ticket
+
+    @asynccontextmanager
+    @async_generator
+    async def lock(
+        self,
+        conversation_id: Text,
+        attempts: int = DEFAULT_LOCK_LIFETIME,
+        wait_time_in_seconds: Union[int, float] = 1,
+    ) -> None:
+        """Acquire lock for `conversation_id` with `ticket`.
+
+        Perform `attempts` with a wait_time_in_seconds of `wait_time_in_seconds` seconds
+         between them before raising a `LockError`.
+        """
+
+        ticket = self.issue_ticket(conversation_id)
+
+        try:
+            # have to use async_generator.yield_() for py 3.5 compatibility
+            await yield_(
+                await self._acquire_lock(
+                    conversation_id, ticket, attempts, wait_time_in_seconds
+                )
+            )
+        finally:
+            self.cleanup(conversation_id, ticket)
+
+    async def _acquire_lock(
+        self, conversation_id: Text, ticket: int, attempts: int, wait: Union[int, float]
+    ) -> TicketLock:
+
+        while attempts > 0:
```

Comment:
why do we need to restrict the attempts? is it not guaranteed that it will acquire the lock eventually?

Reply:
We probably don't, it will just raise a LockError if the lock has expired